File: python/face/api/metodos/sistemas_ecuaciones/eliminacion_simple.py
import numpy as np

from ..numeric_method import NumericMethod
from .matrix_utils import concatenar, intercambiar_filas, sustitucion_regresiva
from .matrix_utils import no_es_invertible, process_params


class EliminacionSimple(NumericMethod):
    def calculate(self, parameters):
        A = parameters["A"]
        b = parameters["b"]

        A, b = process_params(A, b)
        response = self.init_response()

        if no_es_invertible(A):
            response["error"] = "La matriz no es invertible"
            return response

        augmented = eliminacion(A, b)
        # augmented = np.round(augmented, 2)

        A = np.delete(augmented.copy(), len(augmented), 1)
        b = augmented[:, len(augmented)]
        x = sustitucion_regresiva(A, b)

        response["augmented"] = np.round(augmented, 2).tolist()
        response["x"] = x.tolist()

        return response

    def get_description(self):
        return "Retorna la solución a un sistema Ax=b por medio de eliminacion gaussiana"

    def init_response(self):
        response = dict()

        return response


def eliminacion(A, b):
    augmented = concatenar(A, b)
    n = len(augmented)

    for k in range(0, n-1):
        if augmented[k, k] == 0:
            fila = k
            for i in range(k, n):
                if augmented[i, k] != 0:
                    fila = i
                    break
            augmented = intercambiar_filas(augmented, fila, k)

        for i in range(k+1, n):
            mult = augmented[i, k]/augmented[k, k]
            for j in range(k, n+1):
                augmented[i, j] = augmented[i, j] - mult * augmented[k, j]

    return augmented

File: java/face/src/main/java/systemOfEquations/EliminacionSimple.java
package systemOfEquations;

import systemOfEquations.MatrixUtils.*;

import co.edu.eafit.NumericMethod;
import co.edu.eafit.Solution;

public class EliminacionSimple implements NumericMethod {

	@Override
	public Solution calculate(Object... args) {
		double[][] A = (double[][]) args[0];
		double[] b = (double[]) args[1];
		int n = A.length;
		
		if(!MatrixUtils.esInvertible(A)) {
			return new SystemOfEquationsSolution("La matriz no es invertible");
		}
		
		double[][] augmented = MatrixUtils.aumentarMatriz(A, b);
		for (int k 	= 0; k < n-1; k++) {
			if(augmented[k][k] == 0) {
				int fila = k;
				for (int i = k; i < n; i++) {
					if (augmented[i][k] != 0) {
						fila = i;
						break;
					}
				}
				augmented = MatrixUtils.intercambiarFilas(A, fila, k);
			}
			
			for (int i = k+1; i < n; i++) {
				double mult = augmented[i][k] / augmented[k][k];
				for (int j = k; j < n+1; j++) {
					augmented[i][j] = augmented[i][j] - mult * augmented[k][j];
				}
			}
		}
		
		A = getA(augmented);
		b = getB(augmented);
		
		double[] x = MatrixUtils.sustitucionRegresiva(A, b);
				
		return new SystemOfEquationsSolution(x);
		
		
	}
	
	private double[][] getA(double[][] augmented){
		int n = augmented.length;
		double[][] A = new double[n][n];
		
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				A[i][j] = augmented[i][j];
			}
		}
		return A;
		
	}
	
	private double[] getB(double[][] augmented){
		int n = augmented.length;
		double[] b = new double[n];
		
		for (int i = 0; i < n; i++) {
			b[i] = augmented[i][n];
		}
		return b;
	}

	@Override
	public String getName() {
		return "Eliminacion Gaussiana Simple";
	}

}
